### PR TITLE
Add "Insert Silence" as a standalone action out of "Effects" framework

### PR DESCRIPTION
**Pull request by kryksyh**
Resolves: [Effects - Generate Silence](https://github.com/audacity/audacity/issues/7685)	https://github.com/audacity/audacity/issues/7685

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [x] Each commit compiles and runs on my machine without known undesirable changes of behavior

### CODE REVIEW

File: src/trackedit/internal/au3/au3interaction.cpp [hunk 1]
```diff
@@ -1303,6 +1305,82 @@ void Au3Interaction::moveTracksTo(const TrackIdList& trackIds, int to)
...
+        tracks.Add(track);
+
+        trackedit::ITrackeditProjectPtr prj = globalContext()->currentTrackeditProject();
+        prj->notifyAboutTrackAdded(DomConverter::track(track.get()));
+        projectHistory()->pushHistoryState(muse::trc("trackedit", "Insert silence"), muse::trc("trackedit", "Insert silence"));
+        return;
+    }
+
+    for (const auto& trackId : trackIds) {
+        Au3WaveTrack* waveTrack = DomAccessor::findWaveTrack(project, Au3TrackId(trackId));
```
Comment (saintmatthieu) [inline src/trackedit/internal/au3/au3interaction.cpp:1334]: I think you do want to verify the validity of this pointer before proceeding: I've seen this return `nullptr` unexpectedly.

File: src/trackedit/internal/au3/au3interaction.cpp [hunk 2]
```diff
@@ -1303,6 +1305,82 @@ void Au3Interaction::moveTracksTo(const TrackIdList& trackIds, int to)
...
+
+        std::vector<Clip> clipsBefore;
```
Comment (saintmatthieu) [inline src/trackedit/internal/au3/au3interaction.cpp:1336]: Couldn't `ITrackeditProject` have
```cpp
virtual std::vector<Clip> clipList(const TrackId& trackId) const = 0;
virtual muse::async::NotifyList<Clip> clipNotifyList(const TrackId& trackId) const = 0;
```
and then
```cpp
std::vector<au::trackedit::Clip> Au3TrackeditProject::clipList(const au::trackedit::TrackId& trackId) const
{
    const Au3WaveTrack* waveTrack = DomAccessor::findWaveTrack(*m_impl->prj, Au3TrackId(trackId));
    IF_ASSERT_FAILED(waveTrack) {
        return std::vector<au::trackedit::Clip>();
    }

    std::vector<au::trackedit::Clip> clips;
    for (const std::shared_ptr<const Au3WaveClip>& interval : waveTrack->Intervals()) {
        au::trackedit::Clip clip = DomConverter::clip(waveTrack, interval.get());
        clips.push_back(std::move(clip));
    }

    return clips;
}

muse::async::NotifyList<au::trackedit::Clip> Au3TrackeditProject::clipNotifyList(const au::trackedit::TrackId& trackId) const
{
    auto clips = clipList(trackId);
    async::ChangedNotifier<Clip>& notifier = m_clipsChanged[trackId];
    clips.setNotify(notifier.notify());

    return clips;
}
```
?

File: src/trackedit/internal/au3/au3interaction.cpp [hunk 3]
```diff
@@ -1303,6 +1305,82 @@ void Au3Interaction::moveTracksTo(const TrackIdList& trackIds, int to)
...
+        auto rate = ::ProjectRate::Get(project).GetRate();
+
+        auto track = trackFactory.Create(defaultFormat, rate);
+        track->SetName(tracks.MakeUniqueTrackName(Au3WaveTrack::GetDefaultAudioTrackNamePreference()));
+
+        track->InsertSilence(begin, duration);
+
+        tracks.Add(track);
+
+        trackedit::ITrackeditProjectPtr prj = globalContext()->currentTrackeditProject();
```
[the remark this answers is not in the document]
Comment (kryksyh) [reply src/trackedit/internal/au3/au3interaction.cpp:1327]: A typo, thanks

File: src/projectscene/internal/projectsceneuiactions.cpp [hunk 4]
```diff
@@ -128,6 +128,12 @@ const UiActionList ProjectSceneUiActions::m_actions = {
              TranslatableString("action", "Pinned play head"),
              Checkable::Yes
              ),
+    UiAction("insert-silence",
+             au::context::UiCtxAny,
```
[the remark this answers is not in the document]
Comment (kryksyh) [reply src/projectscene/internal/projectsceneuiactions.cpp:132]: It absolutely should!

File: src/trackedit/internal/au3/au3interaction.cpp [hunk 5]
```diff
@@ -1303,6 +1305,82 @@ void Au3Interaction::moveTracksTo(const TrackIdList& trackIds, int to)
     projectHistory()->pushHistoryState("Move track", "Move track");
 }
 
+void Au3Interaction::insertSilence(const TrackIdList trackIds, secs_t begin, secs_t end, secs_t duration)
+{
+    auto prj = globalContext()->currentTrackeditProject();
```
[the remark this answers is not in the document]
Comment (kryksyh) [reply src/trackedit/internal/au3/au3interaction.cpp:1310]: >  why isn't anyone using const for local variables
Please do reject the review whenever you see this.
I love it when some languages (zig comes to mind) give a compiler error if a VARiable is not changed.

File: src/trackedit/internal/au3/au3interaction.cpp (hunk 1 above)
Comment (kryksyh) [reply src/trackedit/internal/au3/au3interaction.cpp:1334]: That's some sloppy code, thanks!

**Review by saintmatthieu — APPROVED** [review]
Native part LGTM !
@Eism , may I leave the review of the QML part to you?

File: src/projectscene/view/common/insertsilencemodel.h [hunk 6]
```diff
@@ -0,0 +1,45 @@
+#ifndef INSERTSILENCEMODEL_H
```
Comment (Eism) [inline src/projectscene/view/common/insertsilencemodel.h:1]: please use `#pragma once`

File: src/projectscene/view/common/insertsilencemodel.h
Comment (Eism) [inline src/projectscene/view/common/insertsilencemodel.h:1]: please add license for new files

File: src/projectscene/view/common/insertsilencemodel.cpp [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+    muse::secs_t begin = selectionController()->dataSelectedStartTime();
+    muse::secs_t end = selectionController()->dataSelectedEndTime();
+
+    m_duration = end - begin;
+    emit durationChanged();
+}
+
+int InsertSilenceModel::sampleRate() const
+{
+    au3::Au3Project* project = reinterpret_cast<au3::Au3Project*>(
```
Comment (Eism) [inline src/projectscene/view/common/insertsilencemodel.cpp:18]: you can use playback()->audioOutput()->sampleRate()

File: src/projectscene/view/common/insertsilencemodel.cpp [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+QString InsertSilenceModel::durationFormat() const
+{
+    return m_durationFormat;
+}
+
+void InsertSilenceModel::setDurationFormat(const QString& newDurationFormat)
+{
+    m_durationFormat = newDurationFormat;
```
Comment (Eism) [inline src/projectscene/view/common/insertsilencemodel.cpp:42]: please add an equality check before it so as not to send an extra change signal

File: src/projectscene/qml/Audacity/ProjectScene/common/InsertSilence.qml [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+    InsertSilenceModel {
+        id: silence
+    }
+
+    Component.onCompleted: {
+        silence.init()
+        silence.durationFormat = root.durationFormat
+        if(silence.duration == 0) {
+            silence.duration = root.duration
```
Comment (Eism) [inline src/projectscene/qml/Audacity/ProjectScene/common/InsertSilence.qml:32]: could you please explain?

File: src/projectscene/qml/Audacity/ProjectScene/common/InsertSilence.qml [hunk 10]
```diff
@@ -0,0 +1,100 @@
+import QtQuick
+import QtQuick.Controls
+import QtQuick.Layouts
+import Muse.UiComponents
+import Audacity.Effects
```
Comment (Eism) [inline src/projectscene/qml/Audacity/ProjectScene/common/InsertSilence.qml:5]: still needed?

File: src/projectscene/qml/Audacity/ProjectScene/common/InsertSilence.qml (hunk 9 above)
Comment (kryksyh) [reply src/projectscene/qml/Audacity/ProjectScene/common/InsertSilence.qml:32]: If there is no selection, set duration to last used, provided by ProjectSceneActionsController

Comment (kryksyh) [reply src/projectscene/qml/Audacity/ProjectScene/common/InsertSilence.qml:32]: Loading configuration can be moved to the model, what do you think?

File: src/trackedit/internal/au3/au3interaction.cpp (hunk 2 above)
Comment (saintmatthieu) [reply src/trackedit/internal/au3/au3interaction.cpp:1339]: @kryksyh what about this comment ?

Comment (kryksyh) [reply src/trackedit/internal/au3/au3interaction.cpp]: Replaced with track change notification.

**Review by Eism — APPROVED** [review]

**Pull request merged**